### PR TITLE
Uplift the invalid_atomic_ordering lint from clippy to rustc

### PR DESCRIPTION
**Pull request by jyn514**
This is mostly just a rebase of https://github.com/rust-lang/rust/pull/79654; I've copy/pasted the text from that PR below.

r? @lcnr since you reviewed the last one, but feel free to reassign.

---

This is an implementation of https://github.com/rust-lang/compiler-team/issues/390.

As mentioned, in general this turns an unconditional runtime panic into a (compile time) lint failure. It has no false positives, and the only false negatives I'm aware of are if `Ordering` isn't specified directly and is comes from an argument/constant/whatever.

As a result of it having no false positives, and the alternative always being strictly wrong, it's on as deny by default. This seems right.

In the [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Uplift.20the.20.60invalid_atomic_ordering.60.20lint.20from.20clippy/near/218483957) @joshtriplett suggested that lang team should FCP this before landing it. Perhaps libs team cares too?

---

Some notes on the code for reviewers / others below

## Changes from clippy

The code is changed from [the implementation in clippy](https://github.com/rust-lang/rust-clippy/blob/68cf94f6a66e47234e3adefc6dfbe806cd6ad164/clippy_lints/src/atomic_ordering.rs) in the following ways:

1. Uses `Symbols` and `rustc_diagnostic_item`s instead of string literals.
    - It's possible I should have just invoked Symbol::intern for some of these instead? Seems better to use symbol, but it did require adding several.
2. The functions are moved to static methods inside the lint struct, as a way to namespace them. 
    - There's a lot of other code in that file — which I picked as the location for this lint because @jyn514 told me that seemed reasonable.
3. Supports unstable AtomicU128/AtomicI128.
    - I did this because it was almost easier to support them than not — not supporting them would have (ideally) required finding a way not to give them a `rustc_diagnostic_item`, which would have complicated an already big macro.
    - These don't have tests since I wasn't sure if/how I should make tests conditional on whether or not the target has the atomic... This is to a certain extent an issue of 64bit atomics too, but 128-bit atomics are much less common. Regardless, the existing tests should be *more* than thorough enough here.
4. Minor changes like:
    - grammar tweaks ("loads cannot have `Release` **and** `AcqRel` ordering" => "loads cannot have `Release` **or** `AcqRel` ordering")
    - function renames (`match_ordering_def_path` => `matches_ordering_def_path`),
    - avoiding clippy-specific helper methods that don't exist in rustc_lint and didn't seem worth adding for this case (for example `cx.struct_span_lint` vs clippy's `span_lint_and_help` helper).

## Potential issues

(This is just about the code in this PR, not conceptual issues with the lint or anything)

1. I'm not sure if I should have used a diagnostic item for `Ordering` and its variants (I couldn't figure out how really, so if I should do this some pointers would be appreciated).
    - It seems possible that failing to do this might possibly mean there are more cases this lint would miss, but I don't really know how `match_def_path` works and if it has any pitfalls like that, so maybe not.

2. I *think* I deprecated the lint in clippy (CC @flip1995 who asked to be notified about clippy changes in the future in [this comment](https://github.com/rust-lang/rust/pull/75671#issuecomment-718731659)) but I'm not sure if I need to do anything else there.
    - I'm kind of hoping CI will catch if I missed anything, since `x.py test src/tools/clippy` fails with a lot of errors with and without my changes (and is probably a nonsense command regardless). Running `cargo test` from src/tools/clippy also fails with unrelated errors that seem like refactorings that didnt update clippy? So, honestly no clue. 

3. I wasn't sure if the description/example I gave good. Hopefully it is. The example is less thorough than the one from clippy here: https://rust-lang.github.io/rust-clippy/master/index.html#invalid_atomic_ordering. Let me know if/how I should change it if it needs changing.

4. It pulls in the `if_chain` crate. This crate was already used in clippy, and seems like it's used elsewhere in rustc, but I'm willing to rewrite it to not use this if needed (I'd prefer not to, all things being equal).

### CODE REVIEW

**Comment by jyn514** [comment]
cc @thomcc

**Comment by jyn514** [comment]
Oh also the clippy deprecation will conflict with https://github.com/rust-lang/rust-clippy/pull/7056; I expect there will be a subtree sync before fcp finishes though so it shouldn't be a big deal.

File: src/tools/clippy/clippy_lints/src/lib.rs
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/lib.rs:1968]: You should also remove the `atomic_ordering.rs` module from Clippy. All the other Clippy changes LGTM.

Comment (Manishearth) [reply src/tools/clippy/clippy_lints/src/lib.rs:1968]: We could do a manual sync

**Comment by jyn514** [comment]
I am not sure why CI is failing :( I think it has something to do with `Acquire` being a Ctor instead of the variant itself, but I'm not sure what's going wrong. Some debugging in https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/How.20do.20I.20get.20the.20ctor.20of.20an.20enum.20variant.3F.

File: src/tools/clippy/clippy_lints/src/lib.rs
Comment (flip1995) [reply src/tools/clippy/clippy_lints/src/lib.rs:1968]: Yeah it also wouldn't hurt to do it a few days earlier. But lets wait for the FCP first.

**Comment by thomcc** [comment]
Hey, thanks for taking this over and carrying it over the finish line. I had struggled a few times with the rebase, since I'm mostly unfamiliar with rustc and clippy, and generally didn't have the time to dedicate to it at the moment.

Still, will be super thrilled to see it land, it will make atomic stuff way easier to teach, if nothing else!

File: compiler/rustc_lint/src/types.rs
Comment (camsteffen) [inline compiler/rustc_lint/src/types.rs:1557]: Should these be static or const?

File: compiler/rustc_lint/src/types.rs
Comment (Manishearth) [reply compiler/rustc_lint/src/types.rs:1467]: Yes, we should IMO

Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1467]: It already exists: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.get_diagnostic_item
It has a warning that it's slower than `is_diagnostic_item` (presumably because it invalids more of the incremental cache). Should I use it anyway?

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs]: Sure, I made them const.

File: library/core/src/sync/atomic.rs
Comment (jyn514) [reply library/core/src/sync/atomic.rs]: Thanks! That fixed all the bugs I was running into earlier :) and it's a lot simpler too.

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1467]: Also, I'm not sure what you're suggesting the new code should look like. Don't I need to check all the types in the array no matter what?

File: src/tools/clippy/clippy_lints/src/lib.rs
Comment (jyn514) [reply src/tools/clippy/clippy_lints/src/lib.rs:1968]: > You should also remove the atomic_ordering.rs module from Clippy.

Thanks, done. I'd actually removed the module already, I just forgot to remove the file :laughing:

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs]: Unfortunately that panics if the method name is the same but the type doesn't match, because it assumes the call has at least 2/3 args.

Comment (jyn514) [reply compiler/rustc_lint/src/types.rs]: Err actually it panics with or without that change; let me look into it.

File: compiler/rustc_lint/src/types.rs
Comment (camsteffen) [reply compiler/rustc_lint/src/types.rs:1467]: > It already exists

That gets a `DefId` from a name. I'm suggesting the opposite.

My second suggestion is to make the parent module a diagnostic item instead of the types and then just check the name of the type.

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs]: @lcnr why was the UI test you suggested valid? Shouldn't inherent impls always take precedence over trait impls? nightly rustc accepts it, but I don't know why :/

**Review by camsteffen — COMMENTED** [review]
A couple more ideas!

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1509]: Thanks, this is a lot cleaner but doesn't fix the ICE.

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1507]: Hmm, this causes a bunch of false negatives:
```
---- [ui] ui/lint/lint-invalid-atomic-ordering-ptr.rs stdout ----

error: ui test compiled successfully!
---- [ui] ui/lint/lint-invalid-atomic-ordering-fetch-update.rs stdout ----

error: ui test compiled successfully!
```

Comment (camsteffen) [reply compiler/rustc_lint/src/types.rs:1507]: Suggested the wrong arg for `type_dependent_def_id` above. Fixed.

Comment (camsteffen) [reply compiler/rustc_lint/src/types.rs:1507]: I'm pretty sure any struct or enum is an Adt.

Comment (camsteffen) [reply compiler/rustc_lint/src/types.rs:1507]: It's the same is [this](https://github.com/rust-lang/rust/pull/84039/files/33963b41631ca3db34731a5fa24bef03d3811905#diff-7b162caf96f4886d206bcbace379df34e22230599c8920968b8dd50dc0d3ee4cR1467)

File: compiler/rustc_lint/src/types.rs
Comment (camsteffen) [inline compiler/rustc_lint/src/types.rs:1554]: ICE is happening at `args[3]`. You should verify the method before getting args. Or `match (method, args)` with slice patterns.

Comment (camsteffen) [reply compiler/rustc_lint/src/types.rs:1554]: [behold](https://github.com/crossbeam-rs/crossbeam-utils/blob/61af67077cb7763033c1e65d079b6438d96bb482/src/atomic/atomic_cell.rs#L212)

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs]: Thanks, that worked! Do you know why the `type_dependent_def_id` for the expr is different from that of the first argument? Shouldn't they both be atomic?

Comment (camsteffen) [reply compiler/rustc_lint/src/types.rs]: See [`ExprKind::MethodCall`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ExprKind.html#variant.MethodCall). It probably just doesn't work for args.

**Comment by lcnr** [comment]
don't have a lot of capacity rn and this is out of cache for me again,

r? @estebank in case you have the time, otherwise I can probably get to this next weekend

**Comment by flip1995** [comment]
Sync PR is up: #84427

**Comment by jyn514** [comment]
I think this is mostly done from an implementation standpoint. @rust-lang/lang does this need an FCP? If so, could someone start one?

**Comment by nikomatsakis** [comment]
Nominating for the @rust-lang/lang meeting, but this seems like a good addition to me.

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by jyn514** [comment]
I think that's the fasted fFCP I've ever seen! :laughing: thank you all!

**Comment by jyn514** [comment]
@estebank this is just waiting on your final review :)

**Comment by jyn514** [comment]
@estebank will you have time to review this sometime soon? If not, do you know a good reviewer?

**Comment by wesleywiser** [comment]
I'll make time to review this week. Thanks for working on this @jyn514 and @thomcc, I'm excited to see this uplifted into rustc!

**Comment by jyn514** [comment]
ping @wesleywiser, do you know when you'll have time to take a look? or if not, do you know a good reviewer?

**Review by wesleywiser — COMMENTED** [review]
Thanks for the ping @jyn514! One question and one small nit. I want to take a closer look at the lint definition tomorrow but overall this lgtm.

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1389]: oops, fixed, thanks!

File: library/core/src/sync/atomic.rs
Comment (jyn514) [reply library/core/src/sync/atomic.rs:202]: Nope! Good catch :)

File: compiler/rustc_lint/Cargo.toml
Comment (wesleywiser) [reply compiler/rustc_lint/Cargo.toml:8]: I would prefer that we try to keep our external dependencies as small as possible and this doesn't seem like a necessary addition to me but we have no official policy saying that, so I don't think it should hold up this PR.

Comment (jyn514) [reply compiler/rustc_lint/Cargo.toml:8]: @wesleywiser if_chain is already in the source tree through clippy, this just adds it to rustc itself. I *could* rewrite it without if_chain, but it would be annoying and IMO a lot uglier.

Comment (wesleywiser) [reply compiler/rustc_lint/Cargo.toml:8]: Yes, that's true but clippy isn't a dependency of rustc so any dependencies they pull in don't actually affect rustc. That's why you had to add it to the tidy allowed set of dependencies. 

It's my personal preference and we already have quite a lot of external dependencies as that list showed so don't worry about it 🙂

**Comment by wesleywiser** [comment]
Seems like a small but consistent performance impact. Looking at the [`incr-unchanged` tests](https://perf.rust-lang.org/detailed-query.html?commit=eca131d06446cc864e89d505ce53e8ed1b5ddd3a&base_commit=d0485c7986e0dac6ffd0207aba56467cb9378d85&benchmark=ripgrep-check&run_name=incr-unchanged), it looks like some of that may be due to the issue I highlighted above where we repeatedly check if each visited method call is to an atomic we are looking for.

File: compiler/rustc_lint/src/types.rs
Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1464]: So, the perf run showed `type_of`, `metadata_decode_entry_type_of`, `diagnostic_items`, `impl_trait_ref`, `metadata_decode_entry_associated_item`, and `metadata_decode_entry_opt_def_kind` being a lot slower. Your caching idea would help with `diagnostic_items`, but I don't think it would help much with any of the others. Instead I'm going to check if the name of the method is one of the four hard-coded function names (load/store/fetch_update/compare_exchange), and only if so do the expensive `type_of` check.

I also have a feeling that the other functions that don't go through `inherent_atomic_method` are slow - only `check_atomic_compare_exchange` uses it.

Comment (jyn514) [reply compiler/rustc_lint/src/types.rs:1464]: Also, I'm not sure how to get a list of the items in a local module :/ item_children panics.

```
error: internal compiler error: compiler/rustc_middle/src/ty/query/mod.rs:285:1: `tcx.item_children(DefId(0:8474 ~ core[c54b]::sync::atomic))` unsupported by its crate; perhaps the `item_children` query was never assigned a provider function
```

**Comment by jyn514** [comment]
diagnostic_items still looks a little expensive. I'll see if I can make the caching work.

**Comment by jyn514** [comment]
Oh hmm in retrospect I think caching won't help much because it still unconditionally checks `is_diagnostic_item` for `check_memory_fence` :/ maybe I could check if `item_name` matches one of `fence` or `compiler_fence` before calling diagnostic item? I think this starting to hit diminishing returns, though.

**Comment by jyn514** [comment]
Hmm, I think this was actually better before. I'll revert to the old approach, without the cache.

**Comment by wesleywiser** [comment]
I doubt we're actually calling any of these functions in the perf benchmarks (or, at least, in any of the contrived/microbenchmarks) so perhaps what we're seeing is just the overhead of adding another late pass itself.

@jyn514 Just to clarify, I assume this version is the same as the one that was benchmarked [here](https://github.com/rust-lang/rust/pull/84039#issuecomment-879568861)?

**Comment by jyn514** [comment]
> I doubt we're actually calling any of these functions in the perf benchmarks (or, at least, in any of the contrived/microbenchmarks) so perhaps what we're seeing is just the overhead of adding another late pass itself.

That's very surprising to me, because the benchmarks are super different before and after I added a fast path. (e.g. there's a 1000% regression in metadata_decode_entry_associated_item in https://perf.rust-lang.org/detailed-query.html?commit=eca131d06446cc864e89d505ce53e8ed1b5ddd3a&base_commit=d0485c7986e0dac6ffd0207aba56467cb9378d85&benchmark=ripgrep-check&run_name=incr-unchanged that disappears in the later runs). Do you mean you doubt that is ever hitting the slow path?

> @jyn514 Just to clarify, I assume this version is the same as the one that was benchmarked here?

Correct, yes.

**Comment by jyn514** [comment]
😕 cargo seems to be generating a different Cargo.lock for me than it is in CI. I don't know why.

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
> Do you mean you doubt that is ever hitting the slow path?

Yes. Most of the benchmarks showing a regression are extremely small (at least, source code wise) and don't call any of the functions we're linting at all.

**Comment by wesleywiser** [comment]
Thanks for working on this @jyn514!

@bors r+

The performance impact seems small overall and the cases impacted the most are extremely small stress-test type benchmarks. Most of the real-world crates don't show much of an impact except for the incr-unchanged case (but even then, it's still quite small).

File: Cargo.lock
Comment (klensy) [inline Cargo.lock:649]: This downgrades crate version.

Comment (jyn514) [reply Cargo.lock]: Good catch, fixed.

**Comment by jyn514** [comment]
@wesleywiser there doesn't seem to be a `// only-target-has-atomic` (or similar) header command. I changed this to only test on amd64 and aarch64 instead, which should test all tier 1 targets. Is that ok? I'm not sure how else I can test this.

File: src/test/ui/lint/lint-invalid-atomic-ordering-int.rs
Comment (wesleywiser) [inline src/test/ui/lint/lint-invalid-atomic-ordering-int.rs:1]: Let's file an issue to track this improvement but this is fine to land.

**Review by wesleywiser — APPROVED** [review]

**Comment by jyn514** [comment]
@wesleywiser do you know why this is still running on android? I have ` // only-x86_64` at the top of the file: https://github.com/rust-lang/rust/blob/768e85bbc5f8001b70bb13c37f0777b4ed232500/src/test/ui/lint/lint-invalid-atomic-ordering-int.rs#L2

**Comment by wesleywiser** [comment]
I'm not sure. It seems like that should work as compiletest is getting invoked with the correct target flag. Maybe try adding `// ignore-android`?

**Comment by camelid** [comment]
triage: @jyn514 it looks like some tests need to be blessed to update line numbers and then this can be re-approved.

@rustbot label: -S-waiting-on-review +S-waiting-on-author

**Comment by jyn514** [comment]
I give up. I have no idea what's wrong and I am not willing to put more time into it.

**Comment by jyn514** [comment]
Oh. For some reason there are like 12 files that all test the same thing, and I only added `// ignore-android` to one of them.

**Comment by wesleywiser** [comment]
Woohoo! Congrats @jyn514 🎉

**Pull request merged**
